File: programs/naiveBayesModel/py-src/naive_bayes.py
import util
import column_table
import class_table


class Naive_bayes:
    def __init__(self):
        self.col_tables  = []
        self.class_table = None

    def train(self, feature_mat, response_vec):
        self.class_table = ClassTable(response_vec)
        num_of_cols = len(feature_mat[0])

        for col_index in range(num_of_cols):
            col = util.extract_column(feature_mat, col_index)
            self.col_tables.append(ColumnTable(col, response_vec))

    # def predict(row):

    def display(self):
        for table in self.col_tables:
            table.display()


class ColumnTable:
    def __init__(self, feature_mat, response_vec):
        self.col_vals   = util.unique(feature_mat)
        self.class_vals = util.unique(response_vec)
        self.counts     = [[0] * len(self.class_vals) for _ in range(len(self.col_vals))]

        for row in range(0, len(feature_mat)):
            col_val   = feature_mat[row]
            class_val = response_vec[row]

            count_row = self.col_vals.index(col_val)
            count_col = self.class_vals.index(class_val)

            self.counts[count_row][count_col] += 1

    def display(self):
        for row in range(len(self.counts)):
            print(str(self.col_vals[row]), ": ", end = "")
            print(self.counts[row])


class ClassTable:
    def __init__(self, response_vec):
        self.class_vals = util.unique(response_vec)
        self.counts     = [0] * len(self.class_vals)

        for item in response_vec:
            class_index = self.class_vals.index(item)
            self.counts[class_index] += 1

File: programs/naiveBayesModel/src/NaiveBayes.java
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;

class NaiveBayes {
    private ColumnTable[] columnTables;
    private ClassTable classTable;

    public void train(String[][] X, String[] Y) {
        int numOfCols = X[0].length;
        columnTables = new ColumnTable[numOfCols];

        for (int colIndex = 0; colIndex < numOfCols; colIndex++) {
            String[] column = Util.extractColumn(X, colIndex);
            columnTables[colIndex] = new ColumnTable(column, Y);
        }

        classTable = new ClassTable(Y);
    }

    public String predict(String[] valuesToPredict) {
        ArrayList<String> classValues = classTable.getClassValues();
        double mostProbableValue = Double.MIN_VALUE;    
        String mostProbableClass = "";

        for (String classValue : classValues) {
            double probability = 1;

            for (int colIndex = 0; colIndex < columnTables.length; colIndex++) {
                probability *= columnTables[colIndex].getProbability(valuesToPredict[colIndex], classValue);
            }

            probability *= classTable.getProbability(classValue);

            System.out.println(classValue + " " + probability);
            if (mostProbableValue < probability) {
                mostProbableValue = probability;
                mostProbableClass = classValue;
            }

        }
        return mostProbableClass;
    }

    public void display() {
        for (ColumnTable table : columnTables) {
            table.display();
        }
    }
}

class ColumnTable{
    private int[][] counts;
    private ArrayList<String> columnValues;
    private ArrayList<String> classValues;

    public ColumnTable(String[] column, String[] Y) {
        columnValues = Util.unique(column);
        classValues = Util.unique(Y);
        counts = new int[columnValues.size()][classValues.size()];

        for (int row = 0; row < column.length; row++) {
            String columnValue = column[row];
            String classValue = Y[row];
            int countRow = columnValues.indexOf(columnValue);
            int countColumn = classValues.indexOf(classValue);

            counts[countRow][countColumn]++;
        }
    }

    public void display() {
        for (int row = 0; row < counts.length; row++) {
            System.out.print(columnValues.get(row) + ": ");
            System.out.println(Arrays.toString(counts[row]));
        }
    }

    public double getProbability(String columnValue, String classValue) {
        int rowIndex = columnValues.indexOf(columnValue);
        int colIndex = classValues.indexOf(classValue);
        int count = counts[rowIndex][colIndex];
        int sum = 0;

        for (int row = 0; row < columnValues.size(); row++) {
            sum += counts[row][colIndex];
        }

        return (double)count / sum;
    }
}

class ClassTable{
    private int[] counts;
    private ArrayList<String> classValues;

    public ClassTable(String[] Y){
        classValues = Util.unique(Y);
        counts = new int[classValues.size()];

        for (String classValue : Y) {
            int classIndex = classValues.indexOf(classValue);
            counts[classIndex]++;
        }
    }

    public ArrayList<String> getClassValues() {
        return classValues;
    }

    public double getProbability(String classValue) {
        int index = classValues.indexOf(classValue);
        int count = counts[index];
        int sum = 0;

        for (int col = 0; col < classValues.size(); col++) {
            sum += counts[col];
        }

        return (double)count / sum;
    }
}
